pyiceberg/io/pyarrow.py

```diff
@@ -1795,15 +1873,19 @@ def write_file(io: FileIO, table_metadata: TableMetadata, tasks: Iterator[WriteT
 
     def write_parquet(task: WriteTask) -> DataFile:
         table_schema = task.schema
-        arrow_table = pa.Table.from_batches(task.record_batches)
+
         # if schema needs to be transformed, use the transformed schema and adjust the arrow table accordingly
         # otherwise use the original schema
         if (sanitized_schema := sanitize_column_names(table_schema)) != table_schema:
             file_schema = sanitized_schema
-            arrow_table = to_requested_schema(requested_schema=file_schema, file_schema=table_schema, table=arrow_table)
+            batches = [
+                to_requested_schema(requested_schema=file_schema, file_schema=table_schema, batch=batch)
+                for batch in task.record_batches
+            ]
         else:
             file_schema = table_schema
-
+            batches = task.record_batches
+        arrow_table = pa.Table.from_batches(batches)
```

Comment:
hmm, looking here, this forced materialization seems to preclude streaming writes, which would you may want if e.g. upserting large amounts of data.  ParquetWriter can be used for streaming writes, so this seems unnecessary?

Reply:
i.e., maybe we could do something like the following?:
```python
    # Check if schema needs to be transformed
    sanitized_schema = sanitize_column_names(table_schema)
    needs_sanitization = sanitized_schema != table_schema

    file_path = f'{table_metadata.location}/data/{task.generate_data_file_path("parquet")}'
    fo = io.new_output(file_path)
    
    with fo.create(overwrite=True) as fos:
        with pq.ParquetWriter(fos, schema=sanitized_schema.as_arrow(), **parquet_writer_kwargs) as writer:
            for batch in task.record_batches:
                if needs_sanitization:
                    sanitized_batch = to_requested_schema(
                        requested_schema=sanitized_schema, file_schema=table_schema, batch=batch
                    )
                else:
                    sanitized_batch = batch
                writer.write_table(pa.Table.from_batches([sanitized_batch]), row_group_size=row_group_size)
 ```

Reply:
Yep I totally agree. I wanted to focus this PR on introducing the reader first, and then work on a subsequent PR to incorporate batches into writes. This just maintains the existing functionality while making use of the refactored `to_requested_schema`

Reply:
ah so the change here is the order of operations. We want to call `to_requested_schema` on each batch first before creating a pyarrow Table from those batches. 

I wonder if we can push `to_requested_schema` up the stack since we already bin-pack batches before passing to `WriteTask`
https://github.com/apache/iceberg-python/blob/a29491af52dc4aff46a325bbaac4a11c2f2bfabc/pyiceberg/table/__init__.py#L2920

Also in this #829, I wanted to introduce schema projection 
https://github.com/apache/iceberg-python/pull/829/files#diff-23e8153e0fd497a9212215bd2067068f3b56fa071770c7ef326db3d3d03cee9bR2904

We can keep `write_parquet` very simple and just handle the writing, and preprocess all the batch/table level operations together.

Reply:
That’s a great suggestion @kevinjqliu - good to see that our work is going to be converging naturally here.

I was hoping to focus on the new read API here and the necessary refactoring in the utility functions, and keep the changes to the write functions to a minimum.

I could incorporate these changes and continue the discussion on updating the write functions in a follow up PR. I think there’s much discussion that are worth continuing on that topic (can we avoid materializing an arrow table and write with record batches)?

Reply:
sgtm, ty!

Reply:
Thank you again for the review!